### PR TITLE
DEX: Uniswap with real tokens + router

### PR DESCRIPTION
**Pull request by Rolaman**
- core contracts
- router contract
- core tasks

### CODE REVIEW

File: contracts.txt [hunk 1]
```diff
@@ -0,0 +1,12 @@
+
```
Comment (0xAleksaOpacic) [inline contracts.txt:1]: WHy do we need this?

File: contracts/UniswapV2Factory.sol [hunk 2]
```diff
@@ -46,29 +45,44 @@ contract UniswapV2Factory is IUniswapV2Factory {
         // assembly {
```
Comment (0xAleksaOpacic) [inline contracts/UniswapV2Factory.sol:45]: Can we remove code which is commented if not used

File: contracts/UniswapV2Pair.sol [hunk 3]
```diff
@@ -224,24 +176,17 @@ contract UniswapV2Pair is NilBase, IUniswapV2Pair {
...
-            if (data.length > 0)
-                IUniswapV2Callee(to).uniswapV2Call(
-                    msg.sender,
-                    amount0Out,
-                    amount1Out,
-                    data
-                );
-            balance0 = TokenLibrary(tokenLib).balanceOf(_token0, address(this));
-            balance1 = TokenLibrary(tokenLib).balanceOf(_token1, address(this));
+            if (amount0Out > 0) _safeTransfer(tokenId0, to, amount0Out); // optimistically transfer tokens
```
Comment (0xAleksaOpacic) [inline contracts/UniswapV2Pair.sol:186]: Now it's not optimistic transfer as we remove flash swaps, so there is not way of reverting it

File: flow.md [hunk 4]
```diff
@@ -0,0 +1,40 @@
+# Flow
```
Comment (0xAleksaOpacic) [inline flow.md:1]: I don't think we need this, as we now have flows in the tasks

File: hardhat.config.ts [hunk 5]
```diff
@@ -68,7 +79,8 @@ const config: NilHardhatUserConfig = {
     },
   },
   walletAddress: process.env.WALLET_ADDR,
-  directTxGasLimit: 500000
+  debug: true,
+  shardId: 1,
```
Comment (0xAleksaOpacic) [inline hardhat.config.ts:83]: shard is already default to 1, we don't need to set

File: ignition/modules/amm.ts [hunk 6]
```diff
@@ -1,7 +1,7 @@
 import { buildModule } from "@nomicfoundation/hardhat-ignition/modules";
```
Comment (0xAleksaOpacic) [inline ignition/modules/amm.ts:1]: Do we need this now if we have a task for it?

File: tasks/deploy_garbage_wallet.ts [hunk 7]
```diff
@@ -0,0 +1,9 @@
+import {task} from "hardhat/config";
```
Comment (0xAleksaOpacic) [inline tasks/deploy_garbage_wallet.ts:1]: We don't need this now with burning?

File: tasks/deploy_token.ts [hunk 8]
```diff
@@ -0,0 +1,12 @@
+import {task} from "hardhat/config";
+
+task("deploy_token", "Deploy token")
+    .setAction(async (taskArgs, hre) => {
+
+        const Token = await hre.ethers.getContractFactory("Token");
+        const contract = await Token.deploy("Token1");
+
+        // await contract.mintCurrency(1000000);
```
Comment (0xAleksaOpacic) [inline tasks/deploy_token.ts:9]: Let' remove comments

File: tasks/faucet.ts [hunk 9]
```diff
@@ -0,0 +1,21 @@
+import {task} from "hardhat/config";
+import {Faucet, Token, Wallet} from "../typechain-types";
+
+task("faucet", "Faucet")
+    .addParam("addr", "Address")
```
Comment (0xAleksaOpacic) [inline tasks/faucet.ts:5]: Why do we need task for faucet now?

File: tasks/flow_1.ts [hunk 10]
```diff
@@ -0,0 +1,15 @@
+import {task} from "hardhat/config";
+
+task("flow_1", "Deploy 2 tokens and factory")
```
Comment (0xAleksaOpacic) [inline tasks/flow_1.ts:3]: Can we somehow make names more description, i would not know what flow_1 means

File: tasks/mint_token.ts [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+import {Token} from "../typechain-types";
+
+task("mint_token", "Mint token")
+    .addParam("token", "Token")
+    .setAction(async (taskArgs, hre) => {
+        const Token = await hre.ethers.getContractFactory("Token");
+        const contract = Token.attach(taskArgs.token) as Token;
+        const result = await contract.mintCurrencyInternal(10000000);
+
+        await sleep(1000);
```
Comment (0xAleksaOpacic) [inline tasks/mint_token.ts:11]: Why do we need await? and sleep?

**Review by 0xAleksaOpacic — CHANGES_REQUESTED** [review]
Let's remove everything we don't need in this repository so we can focus on the stuff we want to show

File: contracts/nil/NilCurrencyBase.sol [hunk 12]
```diff
@@ -0,0 +1,134 @@
+// SPDX-License-Identifier: MIT
```
[the remark this answers is not in the document]
Comment (Rolaman) [reply contracts/nil/NilCurrencyBase.sol:1]: discussed

File: contracts/nil/Nil.sol [hunk 13]
```diff
@@ -0,0 +1,308 @@
+// SPDX-License-Identifier: MIT
```
[the remark this answers is not in the document]
Comment (Rolaman) [reply contracts/nil/Nil.sol:1]: discussed

File: contracts/nil/Faucet.sol [hunk 14]
```diff
@@ -0,0 +1,91 @@
+// SPDX-License-Identifier: MIT
```
[the remark this answers is not in the document]
Comment (Rolaman) [reply contracts/nil/Faucet.sol]: removed

File: contracts/nil/Wallet.sol [hunk 15]
```diff
@@ -0,0 +1,88 @@
+// SPDX-License-Identifier: GPL-3.0
```
[the remark this answers is not in the document]
Comment (Rolaman) [reply contracts/nil/Wallet.sol:1]: discussed

File: contracts.txt (hunk 1 above)
Comment (Rolaman) [reply contracts.txt]: removed

File: contracts/UniswapV2Factory.sol (hunk 2 above)
Comment (Rolaman) [reply contracts/UniswapV2Factory.sol]: removed

File: contracts/UniswapV2Pair.sol (hunk 3 above)
Comment (Rolaman) [reply contracts/UniswapV2Pair.sol]: removed

File: flow.md (hunk 4 above)
Comment (Rolaman) [reply flow.md]: better to remove

File: hardhat.config.ts (hunk 5 above)
Comment (Rolaman) [reply hardhat.config.ts]: agree

File: ignition/modules/amm.ts (hunk 6 above)
Comment (Rolaman) [reply ignition/modules/amm.ts]: removed

File: tasks/flow_1.ts (hunk 10 above)
Comment (Rolaman) [reply tasks/flow_1.ts]: done

File: tasks/mint_token.ts (hunk 11 above)
Comment (Rolaman) [reply tasks/mint_token.ts]: fixed

File: tasks/token_info.ts [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+
+        const Tokens = await hre.ethers.getContractFactory("Token");
+        const token = Tokens.attach(taskArgs.token) as Token;
+
+        const res = await token.getCurrencyName();
+        console.log("name " + JSON.stringify(res));
+
+        const result = await token.getCurrencyId();
+
+        console.log(result);
```
Comment (0xAleksaOpacic) [inline tasks/token_info.ts:16]: Can we improve log to be consistent with name
```suggestion
        console.log("currency id: " + result);
```

File: tasks/pair_reserves.ts [hunk 17]
```diff
@@ -0,0 +1,13 @@
...
+
+task("get_reserves", "Get reserver")
+    .addParam("pair", "Pair")
+    .setAction(async (taskArgs, hre) => {
+
+        const Pair = await hre.ethers.getContractFactory("UniswapV2Pair");
+        const pair = Pair.attach(taskArgs.pair.toLowerCase()) as UniswapV2Pair;
+        const result = await pair.getReserves();
+
+        console.log(result);
```
Comment (0xAleksaOpacic) [inline tasks/pair_reserves.ts:12]: Please improve log here

File: tasks/info.ts [hunk 18]
```diff
@@ -4,12 +4,9 @@ import { HardhatRuntimeEnvironment } from "hardhat/types";
 
 task("info", "Fetch reserves and pair balances")
```
Comment (0xAleksaOpacic) [inline tasks/info.ts:5]: This task is using TokenLib,either delete the task or replace with real implementation

Comment (Rolaman) [reply tasks/info.ts]: removed

**Review by 0xAleksaOpacic — APPROVED** [review]
Lgtm, lets merge this and for fixes and improvements open smaller PR's

**Pull request merged**